cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -205,150 +256,211 @@ std::unique_ptr<column> generate_entry_list_offsets(size_type num_entries,
 }
 
 /**
- * @brief Performs an equality comparison between two entries in a lists column
+ * @brief Performs an equality comparison between two entries in a lists column.
  *
- * For the two elements that are in the same list in the lists column, they will always be
- * considered as different. If they are from the same list and their type is one of floating
- * point types, this functor will return the same comparison result as
- * `cudf::element_equality_comparator`.
+ * For the two elements that are NOT in the same list in the lists column, they will always be
+ * considered as different. If they are from the same list and their type is not floating point,
+ * this functor will return the same comparison result as `cudf::element_equality_comparator`.
  *
  * For floating-point types, entries holding NaN value can be considered as different values or the
- * same value depending on the nans_equal parameter.
+ * same value depending on the `nans_equal` parameter.
  *
- * @tparam Type       The data type of entries
+ * @tparam Type The data type of entries
  * @tparam nans_equal Flag to specify whether NaN entries should be considered as equal value (only
  * applicable for floating-point data column)
  */
-template <class Type, bool nans_equal>
-class list_entry_comparator {
- public:
-  list_entry_comparator(offset_type const* list_offsets,
-                        column_device_view d_view,
-                        null_equality nulls_equal,
-                        bool has_nulls)
-    : list_offsets(list_offsets), d_view{d_view}, nulls_equal{nulls_equal}, has_nulls(has_nulls)
+template <class Type>
+struct column_row_comparator_fn {
+  template <typename T, std::enable_if_t<!cuda::std::is_floating_point_v<T>>* = nullptr>
+  bool __device__ compare(T const& lhs_val, T const& rhs_val) const noexcept
   {
+    return lhs_val == rhs_val;
   }
 
-  template <bool nans_equal_ = nans_equal>
-  std::enable_if_t<cuda::std::is_floating_point_v<Type> and nans_equal_, bool> __device__
-  operator()(size_type i, size_type j) const noexcept
+  template <typename T, std::enable_if_t<cuda::std::is_floating_point_v<T>>* = nullptr>
+  bool __device__ compare(T const& lhs_val, T const& rhs_val) const noexcept
   {
-    // Two entries are not considered for equality if they belong to different lists
+    // If both element(i) and element(j) are NaNs and nans are considered as equal value then this
+    // comparison will return `true`. This is the desired behavior in Pandas.
+    if (nans_equal && std::isnan(lhs_val) && std::isnan(rhs_val)) { return true; }
+
+    // If nans are considered as NOT equal, even both element(i) and element(j) are NaNs this
+    // comparison will still return `false`. This is the desired behavior in Apache Spark.
+    return lhs_val == rhs_val;
```

Comment:
For the purposes of checking for duplicates, are we expecting to be checking for _exact_ duplicates or does potential floating point precision come into play?  If this is the latter, it might useful in the future to examine the overlap between this code and the equality/equivalency checking code in `cudf_test/column_utilities`.

Reply:
This is an interesting question. I thought about this before when firstly implemented it, but since there is no requirement for adding precision control thus I just did an exact comparison. If there is a request for adding precision control then we can go back and add it easily.

Reply:
I don't think this is something we want to entertain. The implication would then be that _all_ libcudf functions that are comparing floating point values for equality should allow controlling the precision. This is just part of how floating point values work and while it makes sense to have that control in testing utilities, putting it in actual compute APIs would be very detrimental to maintenance/performance/testing.